### PR TITLE
Integrate merge table changes

### LINKED ISSUE #6816: Librarian Merge Queue Redesign
### Describe the problem that you'd like solved

EDIT: The designs have changed slightly, see figma.

We'd think we'd like to head in the direction of a Merge Queue Dashboard which looks like this:
https://www.figma.com/file/kOQdKg11suMVZrDSAXTYLc/Merge-Queue?node-id=0%3A1

www figma com_file_kOQdKg11suMVZrDSAXTYLc_Merge-Queue_node-id=0%3A1

Some of this redesign work has already been implemented in this draft: https://github.com/internetarchive/openlibrary/pull/6736

Creating a project board (?) for Librarian Merge Queue and creating issues for the following

## Remaining

1. [ ] Adding `sort` label / link which swaps the sort order from `desc` to `asc`
    * remember sort across requests / pages
    * defaults to `desc` 
2. [ ] Merge Queue: Add UI dropdown to filter by `submitter`
     * remembers submitter across requests / pages
3. [ ] Merge Queue: Add UI dropdown to filter by `reviewer`
    * remembers reviewer across requests / pages
4. [ ] Hide Librarian Merge Queue comments by default in MR cards
    * Have a button which shows # of comments
    * clicking btn hides/shows div
    * Show the _latest_ comment below the title in the MR card (as is pictured in mock up)
5. [ ] Update UI to look more like the above picture

## Completed

- [x] #6556 
- [x] #6927
- [x] Add `mrid` visibly to rows (somewhere) **P2**
- [x] #6812
- [x] #6811 
- [x] #6833
- [x] #6883
- [x] #6915

**Punted**
- [ ] #6996 
- [ ] #6882 - seems too challenging to get ID from infobase internals

### Stakeholders

@seabelis @dcapillae

### LINKED ISSUE #6996: Merge Request Dashboard Chat Redesign
### Evidence / Screenshot (if possible)

It seems like it may not be obvious to some librarians that pressing "enter" in a Merge Request's chat message box will submit it (since the UI hides previous messages).

Here's one thread I expanded where the contributor submitted many of the same message, maybe out of confusion.
openlibrary org_merges_mode_closed_page_4

Maybe **solved** by this redesign: https://www.figma.com/file/kOQdKg11suMVZrDSAXTYLc/Merge-Queue?node-id=0%3A1

### Steps to Reproduce

1. Go to ...
2. Do ...

* Actual:
* Expected:

### Details

- **Logged in (Y/N)?**
- **Browser type/version?**
- **Operating system?**
- **Environment (prod/dev/local)?** prod

### PR DESCRIPTION
**Pull request by jimchamp**
Closes #6816 
Closes #6996

Integrates previous merge table changes.  Refactors JS to use class-based approach.

### Technical

The merge request table is now represented by the `LibrarianQueue` JS class, which references two new classes:
`TableHeader`: Adds listeners to the filters in the table's header
`TableRow`: Provides functionality for a single table row

#### Filtering:
1. Ensure that request filtering works as expected

#### Commenting
1. Clicking the toggle comments button should toggle visibility of the comment preview and the full comments display.  This is true for every usergroup
2. Only the submitter of a request and people with reviewer permissions can add a comment to a request.  The comment input and reply button will only be displayed for the submitter and reviewers.
3. Adding a comment updates the comment preview and the full comments display.

#### As a submitter
1. Should be able to close one's own requests by clicking the red `X` next to the submitter name in the "Opened by" line.
4. The aforementioned red `X` should only appear if the request is "open", and only next to the logged-in submitters name.

#### As a reviewer
1. "Review" button should be present for requests that are "open" and unassigned.
3. Clicking the "Review" button should open the merge UI in a new tab. The button should then be replaced with your username.
5. If a request is "open" and assigned to a reviewer, clicking the red `X` next to the reviewer's name should unassign them.
6. When a reviewer is unassigned, their name is replaced by the "Review" button.
7. "Closed" requests should always display the reviewer with a green checkmark next to their name.

### Screenshot

Screenshot from 2023-07-25 08-38-10

### Next steps
- Replace `<table>` with div-based table (we've drifted away from the `<table>` intended use).
- Add request type somewhere in row ("Author" or "Work")
- Include comment timestamps
- Generalize header filters in the template.

### Stakeholders

@mekarpeles 
@russell-warlow 
@Eds-Dbug

### CODE REVIEW

File: static/css/components/merge-request-table.less [hunk 1]
```diff
@@ -234,3 +137,214 @@
...
+
+  &__comment-display {
+    border: 1px solid @light-grey;
+    background-color: @white;
+    border-radius: 5px;
+    margin-bottom: 6px;
+    overflow: auto;
+    overflow-y: scroll;
+    min-width: 400px;
+    height: 55px;
```
Comment (jimchamp) [inline static/css/components/merge-request-table.less:231]:
```suggestion
    height: 55px;
    resize: vertical;
```

**Comment by jimchamp** [comment]
Slight defect:
I'm seeing that the dot indicators are cropped when the full comments panel is displayed:

Screenshot from 2023-07-26 17-38-28

Will fix while applying suggestions from code review.

File: static/css/components/merge-request-table.less
Comment (cdrini) [inline static/css/components/merge-request-table.less:1]: Noticing some horizontal blank scroll bars.

image

Somewhere an `overflow: scroll` should be changed to an `overflow: auto`

File: openlibrary/templates/merge_queue/merge_queue.html [hunk 2]
```diff
@@ -34,158 +40,166 @@ <h1>$_('Community Edit Requests')</h1>
...
-                $datestr(r['updated'])
-              $elif r.get('reviewer'):
-                $r['reviewer']
-            </td>
-            <td>
+              $# End : Comment section
+
+              $# Start : Submitted by line
+              <span class="mr-details__opened-by">
+                $:_('MR #%(id)s opened %(date)s by <span class="commenter">@%(submitter)s</span>', id=r['id'], date=datestr(r["created"]), submitter=r["submitter"])
```
Comment (cdrini) [inline openlibrary/templates/merge_queue/merge_queue.html:165]: The username here looks like a link but isn't clickable ; can we make it a link ? 
image

File: openlibrary/templates/merge_queue/merge_queue.html [hunk 3]
```diff
@@ -34,158 +40,166 @@ <h1>$_('Community Edit Requests')</h1>
...
+        $ comments = r.get('comments', {}).get('comments', [])
+        $ is_open = r['status'] == 1
+        $ review_url = "%s&mrid=%s" % (r['url'], r['id'])
+        $ is_submitter = username == r['submitter']
+
+        <tr id="mrid-$(r['id'])" class="mr-table-row" data-mrid="$(r['id'])">
+          <td class="mr-info">
+            $# Start : Status dot indicator
+            $ dot_status = 'dot--open' if is_open else 'dot--closed'
+            <div class="dot $dot_status"></div>
```
Comment (cdrini) [inline openlibrary/templates/merge_queue/merge_queue.html:134]: Can we add a `title` here for accessibility ?

**Review by cdrini — APPROVED** [review]
This looks great to me! Playing around with it on testing.openlibrary.org and everything seems to work super slickly! I think we might want to bring this to the librarians/Friday design call/Tuesday to get some real-world UX feedback now everything is implemented! Personally I don't see any UX blockers so ok with merging this and having that feedback result in future tasks :)

Great work @jimchamp ! And @Eds-Dbug on the previous drafts :)

Small bugs found while testing:
- middle clicking the "review" button does not assign me

@mekarpeles do you want to do a quick :+1: for what's on testing as well?

**Review by mekarpeles — APPROVED** [review]
changes lgtm

File: openlibrary/plugins/openlibrary/js/merge-request-table/LibrarianQueue/TableHeader.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+     * Close dropdown menus whose menu button doesn't match a given id.
+     *
+     * @param {string} menuButtonId
+     */
+    closeOtherMenus(menuButtonId) {
+        this.dropMenuButtons.forEach((menuButton) => {
+            if (menuButton.id !== menuButtonId) {
+                menuButton.firstElementChild.classList.add('hidden')
+            }
+        })
```
[the remark this answers is not in the document]
Comment (jimchamp) [reply openlibrary/plugins/openlibrary/js/merge-request-table/LibrarianQueue/TableHeader.js:84]: This function is also used to close open dropdowns, so this won't work.

Do think that the toggling is a bit weird today, but suggesting that any changes occur once a new filter is added.  We recently talked about filtering by request type in a community meeting, and the recent heavy use of the `#open-librarians-g` channel for librarian requests makes me think that new request types and a type filter will be added in the near future.  I expect that DRYing the header template would also happen while that feature is being developed.

**Comment by jimchamp** [comment]
Handled everything but the `closeAllMenus()` function, which can be added when a new filter is added.

@cdrini I'm not sure what a middle click on the "Review" button is supposed to do, but wager that it doesn't do that thing in production today.  Could you open a bug issue with your expectations?

**Comment by jimchamp** [comment]
Let's abstain from merging this until the librarians have had some time to test it out.  If we come to find that some of the data that's missing in this version is vital for librarians' work, we could merge this into the `merge-queue-integration` branch and make the necessary changes there.

I forget: is it feasible to use the merge tools in testing?  Does our LLE caching make it difficult?

**Comment by mekarpeles** [comment]
During next community call we want feedback from @seabelis + @onnotasler.

may be difficult to test author merging on testing.openlibrary.org because of cache.

File: openlibrary/plugins/openlibrary/js/merge-request-table/LibrarianQueue/TableHeader.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (cdrini) [reply openlibrary/plugins/openlibrary/js/merge-request-table/LibrarianQueue/TableHeader.js:84]: Hmm isn't "close open dropdown" the same as "close all dropdowns" ? If I'm understanding you correctly

**Comment by cdrini** [comment]
Oh with regards to middle clicking the "Review" button . Clicking with middle-mouse generally opens something in a new tab in web browsers. That's a native piece of how `<a>` elements work. So in the UI I expect middle clicking to open the review in a new tab _and_ assign me. But yep, it looks like it's an existing issue currently in prod, I'll create an issue :+1:

File: openlibrary/plugins/openlibrary/js/merge-request-table/LibrarianQueue/TableHeader.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (jimchamp) [reply openlibrary/plugins/openlibrary/js/merge-request-table/LibrarianQueue/TableHeader.js:84]: Edit: This function is also used to **open closed** dropdowns.

**Comment by jimchamp** [comment]
> Oh with regards to middle clicking the "Review" button . Clicking with middle-mouse generally opens something in a new tab in web browsers. That's a native piece of how `<a>` elements work. So in the UI I expect middle clicking to open the review in a new tab _and_ assign me. But yep, it looks like it's an existing issue currently in prod, I'll create an issue +1

These links already open a new tab, without the need for a middle click.

**Comment by cdrini** [comment]
Yep! We should support middle-click since it's just the way the web works though ; also unlike clicking it opens the page in a *background* tab, which lets me do stuff like open 10 in new tabs real quick, and then go through them one-by-one!

**Comment by mekarpeles** [comment]
I think based on our conversation + feedback from previous call, this should be good to merge and we can make tweaks later. Thank you @cdrini for reviewing and @jimchamp for shepherding!

**Pull request merged**